### PR TITLE
Generate: Add assisted generation

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Here it is, the PR for assisted generation 🙌 In a nutshell, it uses an assistant model (which should be a smaller model with the same tokenizer) to speed up generation, taking advantage of the reduced need for memory transfers in the main model forward pass. It leverages the same property that makes batched inference faster per token.

Since it is meant to be a reference implementation, the code is meant to be clear and well-commented. If you come across any non-obvious steps, let me know so I can clarify them!

Follow-up steps after this PR:
1. Add support for a `sample` version of assisted generation (many cool apps rely on sampling, including chatbots/assistants)
2. Write a blog post a prepare strong communications about the feature

_________________________________________________________________
To process the potential speedup visually, consider the following script and the two videos. They correspond to greedy search using a 6.9B GPTNeoX model on an nvidia 3090 🚀 

<details>
  <summary>Script</summary>

  ```py
  from transformers import AutoTokenizer, AutoModelForCausalLM, TextStreamer
  import torch
  import time
  
  model_id = "EleutherAI/pythia-6.9b-deduped"
  assistant_id = "EleutherAI/pythia-160m-deduped"
  
  tokenizer = AutoTokenizer.from_pretrained(model_id)
  
  assistant_model = AutoModelForCausalLM.from_pretrained(assistant_id)
  assistant_model = assistant_model.to("cuda")
  
  model_kwargs = {
      "pretrained_model_name_or_path": model_id,
      "device_map": "auto",
      "max_memory": {0: "20GiB", "cpu": "50GiB"},
      "torch_dtype": torch.float16,
  }
  model = AutoModelForCausalLM.from_pretrained(**model_kwargs)
  
  inputs = tokenizer("Here's how to cook a good ramen:", return_tensors="pt").to("cuda")
  
  streamer = TextStreamer(tokenizer=tokenizer)
  
  print("Without assistance:")
  start = time.time()
  model.generate(**inputs, streamer=streamer, max_new_tokens=128)
  print(f"Elapsed time: {time.time() - start:.2f} seconds")
  
  print("With assistance:")
  start = time.time()
  model.generate(**inputs, assistant_model=assistant_model, streamer=streamer, max_new_tokens=128)
  print(f"Elapsed time: {time.time() - start:.2f} seconds")

  ```
</details>

Without assistant             |  With assistant
:-------------------------:|:-------------------------:
 |  

(focus on the speed and the fact that the output is the same, not on the output itself)

### CODE REVIEW

**Comment by gante** [comment]
@amyeroberts @sgugger -- since this PR is a bit more complex than most, I've decided to request a review from you two 🤗

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -1458,6 +1457,66 @@ def test_contrastive_generate_dict_outputs_use_cache(self):
...
+        # - assisted_greedy_search does not support `batch_size > 1`
+
+        for model_class in self.all_generative_model_classes:
+            # won't fix: FSMT and Reformer have a different cache variable type (and format).
+            if any(model_name in model_class.__name__.lower() for model_name in ["fsmt", "reformer"]):
+                return
+            # may fix in the future: the following models fail to pass this test, and need model-specific fixes
+            if any(
+                model_name in model_class.__name__.lower()
+                for model_name in ["bigbirdpegasus", "gptbigcode", "led", "mega", "speech2text"]
```
Comment (gante) [inline tests/generation/test_utils.py:1475]: The fix for these models is non-obvious, so I've decided to prioritize shipping the feature instead of aiming for 100% coverage :)

File: docs/source/en/generation_strategies.mdx [hunk 2]
```diff
@@ -332,3 +332,30 @@ The groups are selected to ensure they are distinct enough compared to the other
...
+<!-- TODO: add link to the blog post about assisted generation when it exists -->
+
+To enable assisted generation, set the `assistant_model` argument with a model.
+
+```python
+>>> from transformers import AutoModelForCausalLM, AutoTokenizer
+
+>>> prompt = "I look forward to"
+>>> checkpoint = "gpt2"
+>>> assistant_checkpoint = "distilgpt2"
```
Comment (sgugger) [inline docs/source/en/generation_strategies.mdx:351]: Would it make more sense to use more recent models here? Also, GPT-2 is already quite fast so maybe a bigger one?

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -1192,14 +1193,18 @@ def generate(
                 on the batch ID `batch_id` and the previously generated tokens `inputs_ids`. This argument is useful
                 for constrained generation conditioned on the prefix, as described in [Autoregressive Entity
                 Retrieval](https://arxiv.org/abs/2010.00904).
-            synced_gpus (`bool`, *optional*):
+            synced_gpus (`bool`, *optional*, defaults to `False`):
```
Comment (sgugger) [inline src/transformers/generation/utils.py:1196]: This should be left as is.

File: src/transformers/generation/utils.py [hunk 4]
```diff
@@ -1145,7 +1145,8 @@ def generate(
         logits_processor: Optional[LogitsProcessorList] = None,
         stopping_criteria: Optional[StoppingCriteriaList] = None,
         prefix_allowed_tokens_fn: Optional[Callable[[int, torch.Tensor], List[int]]] = None,
-        synced_gpus: Optional[bool] = None,
+        synced_gpus: Optional[bool] = False,
```
Comment (sgugger) [inline src/transformers/generation/utils.py:1148]:
```suggestion
        synced_gpus: Optional[bool] = None,
```

**Review by sgugger — APPROVED** [review]
Thanks for working on this! LGTM apart from the change of default of `synced_gpus`.

File: src/transformers/generation/utils.py (hunk 3 above)
Comment (gante) [reply src/transformers/generation/utils.py:1196]: oops, probably some unwanted rebase change! 🙈

File: src/transformers/generation/utils.py [hunk 5]
```diff
@@ -3994,6 +4045,471 @@ def constrained_beam_search(
...
+                        )
+                    else:
+                        decoder_hidden_states = _split_model_outputs(
+                            decoder_hidden_states, outputs.hidden_states, prompt_length, last_matching_idx
+                        )
+
+            # finished sentences should have their next token be a padding token
+            if eos_token_id is not None:
+                if pad_token_id is None:
+                    raise ValueError("If `eos_token_id` is defined, make sure that `pad_token_id` is defined.")
```
Comment (amyeroberts) [inline src/transformers/generation/utils.py:4406]: This check should probably go earlier to fail before starting to generate

**Review by amyeroberts — APPROVED** [review]
Very nice! 🚀  

Just left some very small nits 🤏  Only thing I'd say is the generation method is pretty large, so  might be good to split up - but the numbered comment sections help navigate it a lot 🔍

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -3994,6 +4045,471 @@ def constrained_beam_search(
...
+
+            #  1. Forecast next N tokens using the assistant model. This `for` block can be replaced with a
+            # `.generate()` call if we decide to add `past_key_values` as a possible output of generate, as we
+            # need access to the assistant cache to secure strong speedups.
+            candidate_input_ids = input_ids
+            for _ in range(int(assistant_model.max_assistant_tokens)):
+                # 1.1. use the assistant model to obtain the next candidate logits
+                if "assistant_past_key_values" in model_kwargs:
+                    prev_seq_len = model_kwargs["assistant_past_key_values"][0][0].shape[2]
+                    # `new_token_len` can be 1 or 2 (next token in assistant + last token picked by the larger model)
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py:4221]: I'd rather not add the assert -- the tests will fail if something is wrong here and it will cause slowdowns (which is very undesirable since this is a performace-oriented generation method)

**Comment by gante** [comment]
@amyeroberts regarding splitting up, I totally agree! And not only on this method but on most parts of `GenerationMixin`.  Not only are the functions long, but they reuse a significant part of the logic. I want to address that in the near future, by designing a `.generate()` that can be somehow composed of a sequence of smaller functional blocks. I haven't figured out the deets, but I'd expect that a good implementation would get us better readability, less code duplication, and higher flexibility for HW/model/decoding-specific implementations! 💅 

Before merging, I'm going to double-check that the current code keeps the performance numbers I got a few weeks ago. If everything goes well, it will be merged today 🙏

File: tests/generation/test_utils.py (hunk 1 above)
Comment (keyboardAnt) [reply tests/generation/test_utils.py:1475]: @gante, what issues did they have? (For example, the `gptbigcode` model)

Comment (keyboardAnt) [reply tests/generation/test_utils.py:1475]: @gante, what issues did they have? (For example, the `gptbigcode` model)

**Comment by zhaoyang-star** [comment]
@gante Excellent work! I just dive into the code these days and found that the impl only support batchsize 1. Speculative Decoding have no relative to batchsize. I guess supporting bs >1 will be more hard to impl so you just support bs=1 firstly? Thanks.

**Comment by gante** [comment]
@zhaoyang-star thank you for the kind words :)

Re batch size 1: it was a mix of implementation simplicity and diminishing returns. Since `transformers` works with batched inputs with fixed length, efficiently applying assisted generation/speculative decoding would necessarily mean applying extra logic to realign the tensors (e.g. row 1 might get 5 speculated tokens, but row 2 only gets 2 -- row 2 would need to be left-padded to continue). Moving to [nested tensors](https://pytorch.org/docs/stable/nested.html) will get us rid of this limitation :)

Re implementation differences: the two techniques were developed independently, despite relying on the same principle (saving GPU memory bandwidth with the aid of a smaller model). To put it plainly:
1. Speculative Decoding is better when sampling is active with temperatures above 0.3-0.4 -- it employs a clever mathematical trick to handle decoding mismatches. However, you must define how many tokens you want to fetch from the smaller model.
2. Assisted Generation (our implementation) is better in the other scenarios because it has a dynamic heuristic to decide how many tokens to fetch from the assistant model, based on the assistant hit ratio. This means it can adapt according to the difficulty of the prompt, with additional no user input.

For the record, we will be adding the sampling trick to our implementation soon, so it will be the best of both worlds :)

**Comment by zhaoyang-star** [comment]
@gante Thanks for your reply. 

> Speculative Decoding is better when sampling is active with temperatures above 0.3-0.4 -- it employs a clever mathematical trick to handle decoding mismatches. However, you must define how many tokens you want to fetch from the smaller model.

How to get the conclusion that Speculative Decoding is better when sampling is active with temperatures above 0.3-0.4. And Assisted Generation is better in other scenarios? If the conclusion is right, is it better that we implement both the two methods and decide to execute it according to the vaule of temperature?

BTW, Assisted Generation is much easier to understand than Speculative Decoding. So I perfer to use Assisted Generation.

**Comment by gante** [comment]
@zhaoyang-star The conclusion is empirical, with the `0.3-0.4` being a personal rule of thumb based on my assisted generation tests and the values reported in the speculative decoding paper 🤗 It certainly depends on the model and on the task itself.

After we merge the sampling trick from speculative decoding, calling `assisted_generation` will actually be the best of both worlds -- it will use the sampling trick from speculative decoding AND apply the heuristic to determine the number of candidate tokens from assisted generation, all without additional parameterization!

**Comment by zhaoyang-star** [comment]
@gante Thanks a lot. Can't waiting to try the merged version. I saw https://github.com/huggingface/transformers/pull/27270/ is relative to speculative decoding.

**Comment by Dev-hestabit** [comment]
@gante Have you thought of any solution and approach to implement assisted generation on transformer-nueronx?

**Comment by gante** [comment]
@alvations 👋 

It also works with encoder-decoder models, i.e. models supported by `AutoModelForSeq2SeqLM`. I am definitely unable to list all working cases, but feel free to open an issue if you think should be working and isn't :)
